LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/ExclusiveProcessor.cs

```diff
@@ -0,0 +1,129 @@
+            public static implicit operator TResult(ProcessingOutcome<TResult> outcome) => outcome.Result;
+#pragma warning restore CA1062 // Validate arguments of public methods
+        }
+
+#pragma warning restore CA1034 // Nested types should not be visible
+
+        private sealed record ProcessingTaskOutcome<TResult> :
+            ProcessingOutcome<Task<TResult>>,
+            IProcessingOutcome
+        {
+            public ProcessingTaskOutcome(Task<TResult> task,
+                                         DateTimeOffset submissionTime,
+                                         TimeSpan waitDuration,
+                                         TimeSpan runDuration) :
+                base(task, submissionTime, waitDuration, runDuration)
+            { }
+
+            Task IProcessingOutcome.Task => Result;
+        }
+
+        public async Task<ProcessingOutcome<TResult>> ProcessAsync<TResult>(T processor, Func<Task<TResult>> function)
+        {
+            var outcome = await TryProcessAsync(processor, function);
+            return new ProcessingOutcome<TResult>(await outcome.Result, outcome.SubmissionTime, outcome.WaitDuration, outcome.RunDuration);
+        }
+
+        public async Task<ProcessingOutcome<Task<TResult>>> TryProcessAsync<TResult>(T processor, Func<Task<TResult>> function)
+        {
+            if (function == null) throw new ArgumentNullException(nameof(function));
+
+            var submissionTime = DateTimeOffset.Now;
+
+            lock (this.queue)
+                this.queue.Add(processor);
+
+            Submitted?.Invoke(this, processor);
+
+            while (true)
+            {
+                await this.processingLock.WaitAsync();
+
+                try
+                {
+                    (bool, T) next = default;
+
+                    lock (this.queue)
+                    {
+                        var nextProcessor = this.scheduler(this.queue);
+                        if (!this.comparer.Equals(nextProcessor, processor))
+                            next = (true, nextProcessor);
+                        else
+                            _ = this.queue.Remove(processor);
+                    }
+
+                    if (next is (true, { } someNextProcessor))
+                    {
+                        Interrupted?.Invoke(this, (processor, someNextProcessor));
+                    }
+                    else
+                    {
+                        Processing?.Invoke(this, processor);
+                        var startTime = DateTimeOffset.Now;
+                        var task = await Task.WhenAny(function());
+                        var endTime = DateTimeOffset.Now;
+                        var outcome = new ProcessingTaskOutcome<TResult>(task, submissionTime, startTime - submissionTime, endTime - startTime);
+                        Processed?.Invoke(this, (processor, outcome));
+                        return outcome;
```

Comment:
This is used for logging only and hence I think this should all be UTC.

Reply:
A `DateTimeOffset` always has an offset (even when printed) so there's no ambiguity (like with `DateTime`) whether it's local time or UTC; the former can be converted to the latter any time desired. The decision to use UTC or not can then be done at the time of logging/printing, but I find it's best to capture the value with full fidelity when initialized.

Reply:
As discussed. All default formatting should be UTC. Using UTC now just ensures the default formatting is correct and no one has to take action to ensure we get the right timestamps in the logs.

Reply:
Done with 09b7879898ad752a143fc8beb5f0a6a63465d582.